catalyst_docs/milestone4/milestone4-draft-audit-report.typ

```diff
@@ -0,0 +1,3837 @@
+corrections, for detailed debugging information could significantly improve the
+user experience and reduce troubleshooting time.
+
+=== Resolution
+#v(5pt)
+Pending
+
+#pagebreak()
+
+#v(10pt)
+== ID-U214 Improve Documentation on Optimization Levels
+#v(10pt)
+
+#table(
+  columns: (25pt, 10%, 20%, 20%, 20%),
+  align: center,
+  stroke: none,
+  table.header[][*Level*][*Category*][*Severity*][*Status*],
+  table.cell(fill: rgb("ffff00"))[],
+  [2],
+  [Usability],
+  [Minor],
+  [Pending],
+)
+#v(10pt)
+=== Description
+#v(10pt)
+Currently, there is no clear documentation detailing the different optimization
+levels and the specific constraints that are enabled with each level.  
+Providing
+this information would benefit users of OpShin, as it would give them a better
+understanding of which optimization configuration to choose based on their
+requirement.
+
+=== Recommendation
+#v(5pt)
+The idea behind different Optimization levels(O1,O2,O3) and how the _UPLC_
+differs with each optimization level can be clearly documented with simple
+examples.
+
+=== Resolution
+#v(5pt)
+Pending
+
+#pagebreak()
+
+#v(10pt)
+== ID-U215 Effect of Optimization Levels on Build Output
+#v(10pt)
+
+#table(
+  columns: (25pt, 10%, 20%, 20%, 20%),
+  align: center,
+  stroke: none,
+  table.header[][*Level*][*Category*][*Severity*][*Status*],
+  table.cell(fill: rgb("ffff00"))[],
+  [2],
+  [Usability],
+  [Minor],
+  [Pending],
+)
+#v(10pt)
+=== Description
+#v(10pt)
+When building compiled code, OpShin creates the artifacts based on the default
+optimization level O1, where the conditions set are `constant_folding=False` and
+`remove_dead_code=True`.
+
+As a result, the output _UPLC_ contains more information
+than necessary, and therefore, the generated _CBOR_ is also larger.
+This might increase the script size and makes debugging harder when used in off-chain
+transactions.
+
+=== Recommendation
+#v(5pt)
+When building compiled code, OpShin could use the most aggressive optimizer,
+O3, as the default optimization configuration.
+This would allow users to directly utilize the optimized code without needing to specify any
+optimization levels during the build process.
+
+=== Resolution
+#v(5pt)
+Pending
+
+#pagebreak()
+
+#v(10pt)
+== ID-U216 Improve Error for Out-of-Range Tuple Index
+#v(10pt)
+
+#table(
+  columns: (25pt, 10%, 20%, 20%, 20%),
+  align: center,
+  stroke: none,
+  table.header[][*Level*][*Category*][*Severity*][*Status*],
+  table.cell(fill: rgb("ffff00"))[],
+  [2],
+  [Usability],
+  [Minor],
+  [Pending],
+)
+#v(10pt)
+=== Description
+#v(10pt)
+In `PlutoCompiler.visit_Subscript()` in `compiler.py`, a non user-friendly error is thrown if an out-of-range literal index used when accessing elements of a tuple.
+
+=== Recommendation
+#v(5pt)
+Check out-of-range tuple indexing in `PlutoCompiler.visit_Subscript()` in order to throw a user-friendly error, instead of relying on the error thrown by the Pluthon codebase.
```

Comment:
```suggestion
Check for out-of-range tuple indexing in `PlutoCompiler.visit_Subscript()` to throw a user-friendly error, instead of relying on the error thrown by the Pluthon codebase.
```